core/services/gateway/handlers/webapicapabilities/handler.go

```diff
@@ -1,6 +1,116 @@
 package webapicapabilities
 
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/multierr"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/api"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/config"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/handlers"
+)
+
 const (
 	// NOTE: more methods will go here. HTTP trigger/action/target; etc.
-	MethodWebAPITarget = "web_api_target"
+	MethodWebAPITarget  = "web_api_target"
+	MethodWebAPITrigger = "web_api_trigger"
 )
+
+type handler struct {
+	config         HandlerConfig
+	donConfig      *config.DONConfig
+	don            handlers.DON
+	savedCallbacks map[string]*savedCallback
+	mu             sync.Mutex
+	lggr           logger.Logger
+}
+
+type HandlerConfig struct {
+	MaxAllowedMessageAgeSec uint
+}
+type savedCallback struct {
+	id         string
+	callbackCh chan<- handlers.UserCallbackPayload
+}
+
+var _ handlers.Handler = (*handler)(nil)
+
+func NewWorkflowHandler(handlerConfig json.RawMessage, donConfig *config.DONConfig, don handlers.DON, lggr logger.Logger) (*handler, error) {
+	var cfg HandlerConfig
+	err := json.Unmarshal(handlerConfig, &cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &handler{
+		config:         cfg,
+		donConfig:      donConfig,
+		don:            don,
+		savedCallbacks: make(map[string]*savedCallback),
+		lggr:           lggr.Named("WorkflowHandler." + donConfig.DonId),
+	}, nil
+}
+
+func (d *handler) HandleUserMessage(ctx context.Context, msg *api.Message, callbackCh chan<- handlers.UserCallbackPayload) error {
+	d.mu.Lock()
+	d.savedCallbacks[msg.Body.MessageId] = &savedCallback{msg.Body.MessageId, callbackCh}
+	don := d.don
+	d.mu.Unlock()
+	body := msg.Body
+	var payload TriggerRequestPayload
+	err := json.Unmarshal(body.Payload, &payload)
+	if err != nil {
+		d.lggr.Errorw("error decoding payload", "err", err)
+		callbackCh <- handlers.UserCallbackPayload{Msg: msg, ErrCode: api.UserMessageParseError, ErrMsg: fmt.Sprintf("error decoding payload %s", err.Error())}
+		close(callbackCh)
+		return nil
+	}
+
+	if time.Now().Unix()-int64(d.config.MaxAllowedMessageAgeSec) > payload.Timestamp {
```

Comment:
The user supplies their own timestamp. Does this pose any risk?

I think this check is to protect the user, but may be worth thinking about.

Reply:
From a risk perspective, I don't think there is significant risk.  At the end of the day, they supply the timestamp, and I assume they would timestamp the current time.  So the worst case is that we process old messages that somehow have taken a long time to get through their system.

Or are you worried about parsing the field, injecting bad values?  I believe we'd always get back an int64 so no panics possible.

It does make me wonder if we should use this timestamp as part of the trigger_event_id tho.